django/db/migrations/operations/models.py

```diff
@@ -716,3 +716,103 @@ def database_backwards(self, app_label, schema_editor, from_state, to_state):
 
     def describe(self):
         return "Change managers on %s" % (self.name, )
+
+
+class CreateIndex(Operation):
+    """
+    Creates an index for a model.
+    """
+
+    def __init__(self, model_name, index):
+        self.model_name = model_name
+        self.index = index
+
+    def deconstruct(self):
+        kwargs = {
+            'model_name': self.model_name,
+            'index': self.index,
+        }
+        if hasattr(self, 'name'):
+            kwargs['name'] = self.name
+        return (
+            self.__class__.__name__,
+            [],
+            kwargs
+        )
+
+    def state_forwards(self, app_label, state):
+        model_state = state.models[app_label, self.model_name.lower()]
+        if not hasattr(self.index, 'model'):
+            self.index.model = state.apps.get_model(app_label, self.model_name)
+        if 'indexes' not in model_state.options:
+            model_state.options['indexes'] = []
+        model_state.options['indexes'].append(self.index)
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        to_model = to_state.apps.get_model(app_label, self.model_name)
+        schema_editor.create_index(
+            self.index,
+        )
+
+    def database_backwards(self, app_label, schema_editor, from_state, to_state):
+        from_model = from_state.apps.get_model(app_label, self.model_name)
+        schema_editor.drop_index(
+            self.index,
+        )
+
+    def describe(self):
+        return "Create index on field(s) %s of model %s" % (
+            ', '.join(self.index.fields),
+            self.model_name,
+        )
+
+
+class DropIndex(Operation):
+    """
+    Deletes an index of a model.
+    """
+
+    def __init__(self, model_name, name):
+        self.model_name = model_name
+        self.name = name
```

Comment:
Since this is only a problem when migrations are added by hand, I would prefer @MarkusH's "It's on us to document that you should ALWAYS define the name." approach. If we are to ask the users to explicitly give the index a name why give the pain to all the users, just give it to those who want to add indexes the non-traditional way (via `AddIndex` migration). 
Also the statement "If one writes his own indexes he should really give them explicit names." makes a lot of sense to me. If the user needs to an index using `AddIndex`, I am quite sure he has (or can think of) a name for it.